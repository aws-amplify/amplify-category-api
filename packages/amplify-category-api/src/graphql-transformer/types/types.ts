export type ConstructResourceMeta = {
  rootStack?: StackMeta;
  nestedStack?: StackMeta;
  resourceName: string;
  resourceType: string;
};

export type StackMeta = {
  stackName: string;
  stackType: string;
};
