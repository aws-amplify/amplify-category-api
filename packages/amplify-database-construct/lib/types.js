"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoidHlwZXMuanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi9zcmMvdHlwZXMudHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IiIsInNvdXJjZXNDb250ZW50IjpbIi8qKlxuICogSW5wdXQgcHJvcHMgZm9yIHRoZSBBbXBsaWZ5RGF0YWJhc2UgY29uc3RydWN0LlxuICovXG5leHBvcnQgaW50ZXJmYWNlIEFtcGxpZnlEYXRhYmFzZVByb3BzIHtcbiAgcmVhZG9ubHkgZGVmaW5pdGlvbjogc3RyaW5nO1xufVxuIl19