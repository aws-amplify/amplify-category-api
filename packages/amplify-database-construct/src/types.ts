import { IVpc } from 'aws-cdk-lib/aws-ec2';
import { IDatabaseCluster } from 'aws-cdk-lib/aws-rds';

/**
 * Input props for the AmplifyDatabase construct.
 */
export interface AmplifyDatabaseProps {
  readonly vpc: IVpc;
}

export interface AmplifyDatabaseResources {
  /**
   * The database cluster created by the construct.
   */
  readonly databaseCluster: IDatabaseCluster;
}
