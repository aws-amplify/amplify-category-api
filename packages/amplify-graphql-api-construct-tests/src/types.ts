/**
 * Represents stack outputs generated by Amplify Auth Construct.
 */
export type AuthConstructStackOutputs = {
  authRegion: string;
  userPoolId: string;
  webClientId: string;
  identityPoolId: string;
};
