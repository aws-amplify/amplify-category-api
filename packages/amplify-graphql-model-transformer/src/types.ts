import { SyncConfig } from '@aws-amplify/graphql-transformer-core';

export type ModelTransformerOptions = {
  EnableDeletionProtection?: boolean;
  SyncConfig?: SyncConfig;
};
