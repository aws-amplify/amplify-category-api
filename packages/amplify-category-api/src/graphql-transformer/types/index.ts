export { ConstructResourceMeta, StackMeta } from './types';
export { stacksTypes, getStackMeta, convertToAppsyncResourceObj } from './utils';
