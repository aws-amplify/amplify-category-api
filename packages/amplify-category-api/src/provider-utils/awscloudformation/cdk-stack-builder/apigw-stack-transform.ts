import * as cdk from 'aws-cdk-lib';
import {
  $TSContext,
  AmplifyCategories,
  AmplifyError,
  buildOverrideDir,
  getAmplifyResourceByCategories,
  JSONUtilities,
  PathConstants,
  pathManager,
  stateManager,
  Template,
  writeCFNTemplate,
} from '@aws-amplify/amplify-cli-core';
import * as fs from 'fs-extra';
import * as path from 'path';
import { AmplifyApigwResourceStack, ApigwInputs, CrudOperation, Path } from '.';
import { ApigwInputState } from '../apigw-input-state';
import { ADMIN_QUERIES_NAME } from '../../../category-constants';

export class ApigwStackTransform {
  cliInputs: ApigwInputs;

  resourceTemplateObj: AmplifyApigwResourceStack | undefined;

  cliInputsState: ApigwInputState;

  cfn: Template;

  cfnInputParams: Record<string, any>;

  resourceName: string;

  private _app: cdk.App;

  constructor(context: $TSContext, resourceName: string, cliInputState?: ApigwInputState) {
    this._app = new cdk.App();
    this.resourceName = resourceName;

    // Validate the cli-inputs.json for the resource
    this.cliInputsState = cliInputState ?? new ApigwInputState(context, resourceName);
    this.cliInputs = this.cliInputsState.getCliInputPayload();
    void this.cliInputsState.isCLIInputsValid();
  }

  async transform() {
    let authResourceName: string;

    const pathsWithUserPoolGroups = Object.entries(this.cliInputs.paths).filter(([_, path]) => !!path?.permissions?.groups);

    if (this.resourceName === ADMIN_QUERIES_NAME || pathsWithUserPoolGroups.length > 0) {
      [authResourceName] = getAmplifyResourceByCategories(AmplifyCategories.AUTH).filter(
        (resourceName) => resourceName !== 'userPoolGroups',
      );
    }

    // Generate cloudformation stack from cli-inputs.json
    this.generateStack(authResourceName, pathsWithUserPoolGroups);

    // Modify cloudformation files based on overrides
    await this.applyOverrides();
    // Generate cloudformation stack input params from cli-inputs.json
    this.generateCfnInputParameters();

    // Save generated cloudformation.json and parameters.json files
    await this.saveBuildFiles();
  }

  generateCfnInputParameters() {
    this.cfnInputParams = this.resourceTemplateObj.getCfnParameterValues();
  }

  generateStack(authResourceName?: string, pathsWithUserPoolGroups: [string, Path][] = []) {
    this.resourceTemplateObj = new AmplifyApigwResourceStack(this._app, 'AmplifyApigwResourceStack', this.cliInputs);

    if (authResourceName) {
      const authRoleLogicalId = `auth${authResourceName}UserPoolId`;
      this.resourceTemplateObj.addCfnParameter(
        {
          type: 'String',
          default: authRoleLogicalId,
        },
        authRoleLogicalId,
      );

      const uniqueUserPoolGroupsList = new Set<string>();
      for (const [pathName, path] of pathsWithUserPoolGroups) {
        for (const [groupName, crudOps] of Object.entries(path.permissions.groups)) {
          uniqueUserPoolGroupsList.add(groupName);
          this.resourceTemplateObj.addIamPolicyResourceForUserPoolGroup(
            this.resourceName,
            authRoleLogicalId,
            groupName,
            pathName,
            convertCrudOperationsToCfnPermissions(crudOps),
          );
        }
      }
      Array.from(uniqueUserPoolGroupsList).forEach((userPoolGroupName) => {
        this.resourceTemplateObj.addCfnParameter(
          {
            type: 'String',
            default: `authuserPoolGroups${userPoolGroupName}GroupRole`,
          },
          `authuserPoolGroups${userPoolGroupName}GroupRole`,
        );
      });
    }

    // Add Parameters
    const addedFunctions = new Set();
    for (const path of Object.values(this.cliInputs.paths)) {
      if (!addedFunctions.has(path.lambdaFunction)) {
        addedFunctions.add(path.lambdaFunction);
        this.resourceTemplateObj.addCfnParameter(
          {
            type: 'String',
            default: `function${path.lambdaFunction}Name`,
          },
          `function${path.lambdaFunction}Name`,
        );
        this.resourceTemplateObj.addCfnParameter(
          {
            type: 'String',
            default: `function${path.lambdaFunction}Arn`,
          },
          `function${path.lambdaFunction}Arn`,
        );
      }
    }

    this.resourceTemplateObj.addCfnParameter(
      {
        type: 'String',
      },
      'env',
    );

    // Add conditions
    this.resourceTemplateObj.addCfnCondition(
      {
        expression: cdk.Fn.conditionEquals(cdk.Fn.ref('env'), 'NONE'),
      },
      'ShouldNotCreateEnvResources',
    );

    // Add outputs
    this.resourceTemplateObj.addCfnOutput(
      {
        value: cdk.Fn.join('', [
          'https://',
          cdk.Fn.ref(this.cliInputsState.resourceName),
          '.execute-api.',
          cdk.Fn.ref('AWS::Region'),
          '.amazonaws.com/',
          cdk.Fn.conditionIf('ShouldNotCreateEnvResources', 'Prod', cdk.Fn.ref('env')) as unknown as string,
        ]),
        description: 'Root URL of the API gateway',
      },
      'RootUrl',
    );

    this.resourceTemplateObj.addCfnOutput(
      {
        value: this.resourceName,
        description: 'API Friendly name',
      },
      'ApiName',
    );

    this.resourceTemplateObj.addCfnOutput(
      {
        value: cdk.Fn.ref(this.resourceName),
        description: 'API ID (prefix of API URL)',
      },
      'ApiId',
    );

    // Add resources
    this.resourceName === ADMIN_QUERIES_NAME
      ? this.resourceTemplateObj.generateAdminQueriesStack(this.resourceName, authResourceName)
      : this.resourceTemplateObj.generateStackResources(this.resourceName);
  }

  async applyOverrides(): Promise<void> {
    const backendDir = pathManager.getBackendDirPath();
    const overrideFilePath = pathManager.getResourceDirectoryPath(undefined, AmplifyCategories.API, this.resourceName);
    const overrideJSFilePath = path.join(overrideFilePath, 'build', 'override.js');

    const isBuild = await buildOverrideDir(backendDir, overrideFilePath);

    // skip if packageManager or override.ts not found
    if (isBuild) {
      const { envName } = stateManager.getLocalEnvInfo();
      const { projectName } = stateManager.getProjectConfig();
      const projectInfo = {
        envName,
        projectName,
      };
      try {
        // TODO: Invoke `runOverride` from CLI core once core refactor is done, and
        // this function can become async https://github.com/aws-amplify/amplify-cli/blob/7bc0b5654a585104a537c1a3f9615bd672435b58/packages/amplify-cli-core/src/overrides-manager/override-runner.ts#L4
        // before importing the override file, we should clear the require cache to avoid
        // importing an outdated version of the override file
        // see: https://github.com/nodejs/modules/issues/307
        // and https://stackoverflow.com/questions/9210542/node-js-require-cache-possible-to-invalidate
        delete require.cache[require.resolve(overrideJSFilePath)];
        // eslint-disable-next-line import/no-dynamic-require
        const overrideImport = require(overrideJSFilePath);
        if (overrideImport && overrideImport?.override && typeof overrideImport?.override === 'function') {
          await overrideImport.override(this.resourceTemplateObj as AmplifyApigwResourceStack, projectInfo);
        }
      } catch (err) {
        throw new AmplifyError(
          'InvalidOverrideError',
          {
            message: 'Executing overrides failed.',
            details: err.message,
            resolution: 'There may be runtime errors in your overrides file. If so, fix the errors and try again.',
          },
          err,
        );
      }
    }
  }

  private async saveBuildFiles() {
    if (this.resourceTemplateObj) {
      this.cfn = JSONUtilities.parse(this.resourceTemplateObj.renderCloudFormationTemplate());
    }

    const resourceDirPath = pathManager.getResourceDirectoryPath(undefined, AmplifyCategories.API, this.resourceName);
    fs.ensureDirSync(resourceDirPath);

    const buildDirPath = path.join(resourceDirPath, PathConstants.BuildDirName);
    fs.ensureDirSync(buildDirPath);

    stateManager.setResourceParametersJson(undefined, AmplifyCategories.API, this.resourceName, this.cfnInputParams);

    const cfnFilePath = path.resolve(path.join(buildDirPath, `${this.resourceName}-cloudformation-template.json`));
    return writeCFNTemplate(this.cfn, cfnFilePath);
  }
}

export function convertCrudOperationsToCfnPermissions(crudOps: CrudOperation[]) {
  const opMap: Record<CrudOperation, string[]> = {
    [CrudOperation.CREATE]: ['/POST'],
    [CrudOperation.READ]: ['/GET'],
    [CrudOperation.UPDATE]: ['/PUT', '/PATCH'],
    [CrudOperation.DELETE]: ['/DELETE'],
  };
  return crudOps.flatMap((op) => opMap[op]);
}
