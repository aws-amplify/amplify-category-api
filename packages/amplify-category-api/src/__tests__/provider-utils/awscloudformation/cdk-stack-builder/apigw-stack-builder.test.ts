import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { AmplifyApigwResourceStack } from '../../../../provider-utils/awscloudformation/cdk-stack-builder/apigw-stack-builder';
import { CrudOperation, PermissionSetting } from '../../../../provider-utils/awscloudformation/cdk-stack-builder/types';
import { convertCrudOperationsToCfnPermissions } from '../../../../provider-utils/awscloudformation/cdk-stack-builder/apigw-stack-transform';

describe('AmplifyApigwResourceStack', () => {
  test('generateStackResources should synthesize the way we expected', () => {
    const app = new cdk.App();
    const amplifyApigwStack = new AmplifyApigwResourceStack(app, 'amplifyapigwstack', {
      version: 1,
      paths: {
        '/path': {
          lambdaFunction: 'lambdaFunction',
          permissions: {
            setting: PermissionSetting.OPEN,
          },
        },
      },
    });
    amplifyApigwStack.generateStackResources('myapi');
    const template = Template.fromStack(amplifyApigwStack);
    template.hasResourceProperties('AWS::ApiGateway::GatewayResponse', {
      ResponseType: 'DEFAULT_4XX',
      ResponseParameters: {
        'gatewayresponse.header.Access-Control-Allow-Origin': "'*'",
        'gatewayresponse.header.Access-Control-Allow-Headers': "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'",
        'gatewayresponse.header.Access-Control-Allow-Methods': "'DELETE,GET,HEAD,OPTIONS,PATCH,POST,PUT'",
        'gatewayresponse.header.Access-Control-Expose-Headers': "'Date,X-Amzn-ErrorType'",
      },
      RestApiId: {
        Ref: 'myapi',
      },
    });
    template.hasResourceProperties('AWS::ApiGateway::GatewayResponse', {
      ResponseType: 'DEFAULT_5XX',
      ResponseParameters: {
        'gatewayresponse.header.Access-Control-Allow-Origin': "'*'",
        'gatewayresponse.header.Access-Control-Allow-Headers': "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'",
        'gatewayresponse.header.Access-Control-Allow-Methods': "'DELETE,GET,HEAD,OPTIONS,PATCH,POST,PUT'",
        'gatewayresponse.header.Access-Control-Expose-Headers': "'Date,X-Amzn-ErrorType'",
      },
      RestApiId: {
        Ref: 'myapi',
      },
    });
  });
  test('addIamPolicyResourceForUserPoolGroup should generate correct IAM policies when {} is used in path', () => {
    const app = new cdk.App();
    const mockResourceName = 'mockRestAPIName';
    const mockAuthRoleLogicalId = 'mockAuthRoleLogicalId';
    const cliInputs = {
      version: 1,
      paths: {
        '/book/{isbn}': {
          lambdaFunction: 'lambdaFunction1',
          permissions: {
            setting: PermissionSetting.PRIVATE,
            groups: {
              admin: [CrudOperation.CREATE, CrudOperation.UPDATE, CrudOperation.READ, CrudOperation.DELETE],
              member: [CrudOperation.READ],
            },
          },
        },
        '/items/{id}/foobar': {
          lambdaFunction: 'lambdaFunction2',
          permissions: {
            setting: PermissionSetting.PRIVATE,
            groups: {
              admin: [CrudOperation.CREATE, CrudOperation.UPDATE, CrudOperation.READ, CrudOperation.DELETE],
              member: [CrudOperation.READ],
            },
          },
        },
      },
    };
    const amplifyApigwStack = new AmplifyApigwResourceStack(app, 'amplifyapigwstack', cliInputs);
    const pathsWithUserPoolGroups = Object.entries(cliInputs.paths).filter(([_, path]) => !!path?.permissions?.groups);
    for (const [pathName, path] of pathsWithUserPoolGroups) {
      for (const [groupName, crudOps] of Object.entries(path.permissions.groups)) {
        amplifyApigwStack.addIamPolicyResourceForUserPoolGroup(
          mockResourceName,
          mockAuthRoleLogicalId,
          groupName,
          pathName,
          convertCrudOperationsToCfnPermissions(crudOps),
        );
      }
    }

    const template = Template.fromStack(amplifyApigwStack);
    expect(template.findResources('AWS::IAM::Policy')).toMatchSnapshot();
  });
});
