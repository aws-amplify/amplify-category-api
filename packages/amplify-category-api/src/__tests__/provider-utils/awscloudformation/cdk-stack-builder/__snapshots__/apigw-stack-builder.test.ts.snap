// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmplifyApigwResourceStack addIamPolicyResourceForUserPoolGroup should generate correct IAM policies when {} is used in path 1`] = `
Object {
  "adminGroupbookisbnPolicy": Object {
    "Properties": Object {
      "PolicyDocument": Object {
        "Statement": Array [
          Object {
            "Action": "execute-api:Invoke",
            "Effect": "Allow",
            "Resource": Array [
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/POST/book/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/PUT/book/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/PATCH/book/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/GET/book/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/DELETE/book/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/POST/book/*/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/PUT/book/*/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/PATCH/book/*/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/GET/book/*/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/DELETE/book/*/*",
                  ],
                ],
              },
            ],
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "mockRestAPIName-bookisbn-admin-group-policy",
      "Roles": Array [
        Object {
          "Fn::Join": Array [
            "-",
            Array [
              Object {
                "Ref": "mockAuthRoleLogicalId",
              },
              "adminGroupRole",
            ],
          ],
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
  "adminGroupitemsidfoobarPolicy": Object {
    "Properties": Object {
      "PolicyDocument": Object {
        "Statement": Array [
          Object {
            "Action": "execute-api:Invoke",
            "Effect": "Allow",
            "Resource": Array [
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/POST/items/*/foobar",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/PUT/items/*/foobar",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/PATCH/items/*/foobar",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/GET/items/*/foobar",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/DELETE/items/*/foobar",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/POST/items/*/foobar/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/PUT/items/*/foobar/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/PATCH/items/*/foobar/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/GET/items/*/foobar/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/DELETE/items/*/foobar/*",
                  ],
                ],
              },
            ],
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "mockRestAPIName-itemsidfoobar-admin-group-policy",
      "Roles": Array [
        Object {
          "Fn::Join": Array [
            "-",
            Array [
              Object {
                "Ref": "mockAuthRoleLogicalId",
              },
              "adminGroupRole",
            ],
          ],
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
  "memberGroupbookisbnPolicy": Object {
    "Properties": Object {
      "PolicyDocument": Object {
        "Statement": Array [
          Object {
            "Action": "execute-api:Invoke",
            "Effect": "Allow",
            "Resource": Array [
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/GET/book/*",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/GET/book/*/*",
                  ],
                ],
              },
            ],
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "mockRestAPIName-bookisbn-member-group-policy",
      "Roles": Array [
        Object {
          "Fn::Join": Array [
            "-",
            Array [
              Object {
                "Ref": "mockAuthRoleLogicalId",
              },
              "memberGroupRole",
            ],
          ],
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
  "memberGroupitemsidfoobarPolicy": Object {
    "Properties": Object {
      "PolicyDocument": Object {
        "Statement": Array [
          Object {
            "Action": "execute-api:Invoke",
            "Effect": "Allow",
            "Resource": Array [
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/GET/items/*/foobar",
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "mockRestAPIName",
                    },
                    "/",
                    Object {
                      "Fn::If": Array [
                        "ShouldNotCreateEnvResources",
                        "Prod",
                        Object {
                          "Ref": "env",
                        },
                      ],
                    },
                    "/GET/items/*/foobar/*",
                  ],
                ],
              },
            ],
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "mockRestAPIName-itemsidfoobar-member-group-policy",
      "Roles": Array [
        Object {
          "Fn::Join": Array [
            "-",
            Array [
              Object {
                "Ref": "mockAuthRoleLogicalId",
              },
              "memberGroupRole",
            ],
          ],
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
}
`;
